feat: substra example (#219)
owkin/FLamby: integration/Substra/README.md

# SubstraFL example on FLamby TCGA BRCA dataset

> This example illustrate the basic usage of SubstraFL, and propose a model training by Federated Learning

Reviewer: illustrates/proposes

> using de Federated Average strategy on the TCGA BRCA dataset, using FLamby.

Reviewer: the Federated Averaging strategy on the TCGA BRCA dataset with FLamby

# Substra

[Substra](https://docs.substra.org/en/latest/) is an open source federated learning framework. It provides a flexible python interface and a web app to run federated learning training at scale.

Substra main usage is a production one. It has already been deployed and used by hospitals and biotech companies (see the [MELLODDY](https://www.melloddy.eu/) project for instance). Yet Substra can also be used on a single machine on a virtually splitted dataset for two use cases:

- debugging code before launching experiments on a real network
- performing FL simulations

Substra was created by [Owkin](https://owkin.com/) and is now hosted by the [Linux Foundation for AI and Data](https://lfaidata.foundation/).

# SubstraFL example on FLamby TCGA BRCA dataset

> This example illustrate the basic usage of SubstraFL, and propose a model training by Federated Learning using de Federated Average strategy on the TCGA BRCA dataset.

Reviewer: ```suggestion
This example illustrate the basic usage of SubstraFL, and propose a model training by Federated Learning using the Federated Averaging strategy on the TCGA BRCA dataset.
```

> The objective of this example is to launch a *federated learning* experiment on the six centers provided on Flamby (called organization on Substra), using the **FedAvg strategy** on the **baseline model**.

Reviewer: ```suggestion
The objective of this example is to launch a *federated learning* experiment on the six centers provided on Flamby (called organizations in Substra), using the **FedAvg strategy** on the **baseline model**.
```

> This example runs in local mode, no need to deploy a Substra platform to run it.

Reviewer: ```suggestion
This example runs in local mode in python runtime, but can also be run on a deployed substra network.
```

Author: As flamby is not easily instalable with pip for now, the example does not support the docker mode nor the deployed mode. We might want to update the example later for both these mode?

This example is designed to work in local subprocess mode, and not in local Docker mode. As FLamby is a dependency for both the metric and the algorithm, it would need to be installed in editable mode in each container, which has not been added to the example yet.

## Requirements

> Please ensure to have all the libraries installed, a *requirements.txt* file is included in the example, where you can run the command: `pip install -r requirements.txt` to install them.

Reviewer: ```suggestion
Please ensure to have, in addition to flamby, all the libraries needed for this example installed, a *requirements.txt* file is included in the example, this way you can run the command: `pip install -r requirements.txt` to install them.
```

## Objective

This example runs a federated training on all 6 centers.

This example shows how to interface Substra with Flamby. The modifications to use this example on other dataset are minimal and have to be made on the openers files, the metric file and the main script.
This example runs in local mode, simulating a **federated learning** experiment.

## Run the example

> To run the example, launch `jupyter notebook` and run the cells of the `plot_substrafl_flamby.ipynb`

Reviewer: Can we ditch the notebook in favor of markdown files to keep homogeneity ?

Author: Do we want to keep both like febiomed or only keep the markdown ?

# Substra

> [Substra](https://docs.substra.org/en/latest/) is an open source federated learning framework. It provides a flexible python interface and a web app to run federated learning training at scale.

Reviewer 2: ```suggestion
[Substra](https://docs.substra.org/) is an open source federated learning framework. It provides a flexible python interface and a web app to run federated learning training at scale.
```

`latest` does not match any release, let's not redirect users to it
I expect this link to be functional in the next week

Author: Yes, I pointed it out in the PR description. I'm waiting for the release to fix the doc link

Substra main usage is a production one. It has already been deployed and used by hospitals and biotech companies (see the [MELLODDY](https://www.melloddy.eu/) project for instance). Yet Substra can also be used on a single machine on a virtually splitted dataset for two use cases:

> debugging code before launching experiments on a real network
> 
> performing FL simulations

Reviewer 2: ```suggestion
- debugging code before launching experiments on a real network
- performing FL simulations
```

Substra was created by [Owkin](https://owkin.com/) and is now hosted by the [Linux Foundation for AI and Data](https://lfaidata.foundation/).

# SubstraFL example on FLamby TCGA BRCA dataset

This example illustrate the basic usage of SubstraFL, and propose a model training by Federated Learning
using de Federated Average strategy on the TCGA BRCA dataset.

The objective of this example is to launch a *federated learning* experiment on the six centers provided on Flamby
(called organization on Substra), using the **FedAvg strategy** on the **baseline model**.

> This example does not use the deployed platform of Substra and will run in local mode.

Reviewer 2: ```suggestion
This example runs in local mode, no need to deploy a Substra platform to run it.
```

## Requirements

- To run this example locally, please make sure to download and unzip in the same directory as this example the
    assets needed to run it:

    Please ensure to have all the libraries installed, a *requirements.txt* file is included in the zip file, where

>     you can run the command: `pip install -r requirements.txt` to install them.

Reviewer 2: no zip folder here, you can directly say to do `pip install -r requirements.txt`

Author: Copy past too fast, thanks for the catch !

## Objective

> This example will run a federated training on all 6 centers.

Reviewer 2: ```suggestion
This example runs a federated training on all 6 centers.
```
